src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.cs

```diff
@@ -0,0 +1,228 @@
+                    bool hasReference = parameterAndUsageKvp.Value;
+
+                    ReportUnusedParameterDiagnostic(parameter, hasReference, context.ReportDiagnostic, context.Options, context.CancellationToken);
+                }
+            }
+
+            private void ReportUnusedParameterDiagnostic(
+                IParameterSymbol parameter,
+                bool hasReference,
+                Action<Diagnostic> reportDiagnostic,
+                AnalyzerOptions analyzerOptions,
+                CancellationToken cancellationToken)
+            {
+                if (!IsUnusedParameterCandidate(parameter))
+                {
+                    return;
+                }
+
+                var location = parameter.Locations[0];
+                var optionSet = analyzerOptions.GetDocumentOptionSetAsync(location.SourceTree, cancellationToken).GetAwaiter().GetResult();
+                if (optionSet == null)
+                {
+                    return;
+                }
+
+                var option = optionSet.GetOption(CodeStyleOptions.UnusedParameters, parameter.Language);
+                if (option.Notification.Severity == ReportDiagnostic.Suppress ||
+                    !ShouldReportUnusedParameters(parameter.ContainingSymbol, option.Value, option.Notification.Severity))
+                {
+                    return;
+                }
+
+                var diagnostic = DiagnosticHelper.CreateWithMessage(s_unusedParameterRule, location,
+                    option.Notification.Severity, additionalLocations: null, properties: null, message: GetMessage());
+                reportDiagnostic(diagnostic);
+
+                return;
+
+                // Local functions.
+                LocalizableString GetMessage()
+                {
+                    LocalizableString messageFormat;
+                    if (parameter.ContainingSymbol.HasPublicResultantVisibility() &&
+                        !parameter.ContainingSymbol.IsLocalFunction())
+                    {
+                        messageFormat = hasReference
+                            ? FeaturesResources.Remove_unused_parameter_0_if_it_is_not_part_of_a_shipped_public_API_its_initial_value_is_never_used
+                            : FeaturesResources.Remove_unused_parameter_0_if_it_is_not_part_of_a_shipped_public_API;
+                    }
+                    else if (hasReference)
+                    {
+                        messageFormat = FeaturesResources.Remove_unused_parameter_0_its_initial_value_is_never_used;
+                    }
+                    else
+                    {
+                        messageFormat = s_unusedParameterRule.MessageFormat;
+                    }
+
+                    return new DiagnosticHelper.LocalizableStringWithArguments(messageFormat, parameter.Name);
+                }
+            }
+
+            private static IEnumerable<INamedTypeSymbol> GetAttributesForMethodsToIgnore(Compilation compilation)
```

Comment:
is this called multiple times?  should this maybe not be an ienumerable?  also, consider just yielding the values from an inner helper and calling .WhereNotNull on hte result.

Reply:
This is only called once, I can change to use WhereNotNull

---
In reply to: [234868788](https://github.com/dotnet/roslyn/pull/30777#discussion_r234868788) [](ancestors = 234868788)